### DISCUSSION TITLE
Literal format for the Map type

### DISCUSSION THREAD

**Discussion by jdunkerley** [enso-org/enso#6167] [Work In Progress - Engine]
As we now have a built-in Map type we would like to add a literal format to create a map instance. 

There appears to have been old support within the old parser for a syntax:
https://github.com/enso-org/enso/blob/d1c52fef0b935744dc880dbcb705f2a3be857ad8/engine/runtime/src/test/scala/org/enso/compiler/test/pass/analyse/DataflowAnalysisTest.scala#L1295

The format appears to have been: `{ x := 1; b := 2 }`.

Readding support to this for the parser would be the first step, and then making the engine produce a Map from it.

For example:
- `{ x := 1; b := 2}` would be translated to `Map.empty.insert a 1 . insert b 2`
- `{ "key" := 1, "value" := 2}` would be translated to `Map.empty.insert "key" 1 . insert "value" 2`

A ticket for the parser has been added by @kazcw: #6162 

Is everyone happy on adding this back in? 
I would suggest switching to `,` as the delimiter rather than `;` as appears to be the case in the old parser.

cc: @hubertp @Akirathan @JaroslavTulach @wdanilo @radeusgd @GregoryTravis

**Comment by wdanilo** [comment]
The old format was accidental and we never designed it properly. I believe that the old syntax does not fit the current Enso language design. I do not know yet what design is correct though. I think something like `{a = 1, b = 2, c = 3}` is already better than the one above, but what I don't like here is the need of using curly braces, even with multi-line definition, which is something new in Enso syntax. Alternatively, we could do something like (`map` is a macro here): `v1 = map (a = 1, b = 2, c = 3)` which could be used as:

```
v1 = map
    a = 1
    b = 2
    c = 3
```

Anyway, this needs IMO a little bit more discussion and brainstorming. I'd love to hear what you think about these ideas.

**Comment by jdunkerley** [comment]
The curly brace felt natural to me as it is very similar to python (although python uses `:`) and JavaScript. 
Agree just having an `=` is cleaner than the `:=` above.

We use square brackets exclusively for Vector literals so having a similar unique marker for brackets feels in keeping.

The macro syntax would be a new thing we haven't had before (or at least not such that an average user would know).

**Reply by wdanilo** [reply]
Regarding the macro syntax - it's not very magical to the user. We can simply tell that "`map` is a function which accepts as many named arguments as you want". And basically that's it, as the above syntax works for such functions out of the box.

The argument with us having array literal (`[...]`) is a good one, unfortunately. Unfortunately, because the array literal has the same problems as I described - the multi-line form feels "not like Enso", as it requires commas and closing bracket, which our syntax normally does not. Of course, we could create similar macro:

```
v = array 1 2 3
v2 = array
    1 + 2
    3 * 7
    6 - 9
```

I'm not suggesting to do it though, just spawning ideas / thoughts. 

So, to sum this up, we have 2 major syntax proposals here:

<table>
<tr>
<td> Operation </td> <td> Requiring closing delimiters </td> <td> Not requiring closing delimiters </td>
</tr>

<tr>
<td> In-line array definition </td>
<td> 

`v = [1, 2, 3]`

</td>
<td> 

`v = array 1 2 3` 

</td>
</tr>

<tr>
<td> In-line map definition </td>
<td> 

`v = {a=1, b=2, c=3}`

</td>
<td> 

`v = map a=1 b=2 c=3` 

</td>
</tr>


<tr>
<td> Multi-line array definition </td>
<td> 

```
v = [
    1 + 2, 
    3 * 4, 
    9 / 7
]
```

</td>
<td> 

```
v = array 
    1 + 2
    3 * 4 
    9 / 7
``` 

<tr>
<td> Multi-line map definition </td>
<td> 

```
v = {
    a = 1, 
    b = 2, 
    c = 3
}
```

</td>
<td> 

```
v = map 
    a = 1 
    b = 2 
    c = 3
``` 

</td>
</tr>


Honestly, I believe the "Not requiring closing delimiters" is WAY more Enso-like and I feel it suits the language better. I'm still not convinced though, so please read it as further spawning of thoughts here.

**Reply by JaroslavTulach** [reply]
How will the parser recognize that:
```
1 + 2
3 * 4
9 / 7
```
is a literal and not a block of statements? How will the parser recognize that
```
a = 1
b = 2
c = 3
```
is a literal and not a block of statements? There is `map` function on `Vector`, `Array`. How will the parser know that `map` means to create a new map, and not invoke the `map` function?

**Reply by wdanilo** [reply]
Currently parser handles that already. If you write 

```
v = map 
    a = 1 
    b = 2 
    c = 3
```

It is interpreted as 

```
v = map (a = 1) (b = 2) (c = 3)
```

Where `(a = 1)` is a named argument to the function.

Regarding the question how parser would recognize that `map` means a new map, not a function, the answer is that it will not. In the proposal above, `map` is just a function that takes arbitrary number of arguments. It is possible to create such functions using macros in Enso now. An alternative to macros could be something that is pretty common in Haskell, a function builder. So, basically `map: Map_Builder` and it is just a struct, which behaves like an empty `Map`, but if you provide it with an argument, it should consume it and add it to the underlying map. This can be done by defining a `call` method on `Map_Builder`.

**Reply by JaroslavTulach** [reply]
Currently
```haskell
x (a=1) (b=2) (c=3) = a + b + c

main = x
    a = 5
    b = 7
    c = 9
```
yields an error:
```
$ enso --run x.enso
In module x:
Compiler encountered errors:
x.enso[5:5-5:5]: The name `a` could not be found.
x.enso[6:5-6:5]: The name `b` could not be found.
x.enso[7:5-7:5]: The name `c` could not be found.
Aborting due to 3 errors and 0 warnings.
```

**Reply by jdunkerley** [reply]
For the in-line definitions I feel the `[]` and `{}` will feel more common to users. 
The `[]` syntax for vectors and lists is so common across a very wide range of languages.
For the data prep side I think these should be included.

I don't have any objections to having the enso syntax as well but feel the brackets syntax will be the most natural to users.

**Reply by wdanilo** [reply]
@JaroslavTulach that's bad. Hmm, the new parser should support that. Could you check out the AST you've got from that? CC @kazcw, are we supporting this form?

**Reply by kazcw** [reply]
The body of `main` parses as:
```clojure
(ArgumentBlockApplication (Ident x) #(
 (OprApp (Ident a) (Ok "=") (Number () "5" ()))
 (OprApp (Ident b) (Ok "=") (Number () "7" ()))
 (OprApp (Ident c) (Ok "=") (Number () "9" ()))))
```
This representation is awfully low-level, but not inaccurate. (Some day I'd like to introduce an `enum ArgumentBlockValue`--a bunch of things are `Tree` that I'd prefer to restrict.).

It looks like translation doesn't support named-arguments in argument-block-applications. Translation work was driven by parsing the `.enso` corpus and passing the stdlib tests; this case never came up, and I didn't notice support was missing.

**Reply by JaroslavTulach** [reply]
> This representation is ... low-level, but not inaccurate

True, I can see how to convert it into `IR`.

> translation doesn't support named-arguments in argument-block-applications

The remaining question is, what's the key differentiator from:
```haskell
main = x
    a = 5
    b = 7
    a * b
```
e.g. a block of statements (like in the else branch of `if_then_else`?). When one is supposed to treat `ArgumentBlockApplication` as set of named arguments and when one is supposed to treat it as a sequence of statements?

**Reply by JaroslavTulach** [reply]
> `map` is just a function that takes arbitrary number of arguments.

I don't think we have any support for that in the IR and after.

> It is possible to create such functions using macros in Enso now. 

Maybe, but the rest of the pipeline is unlikely ready for them. We were hoping for a focused, small change to support map literals - not a support for a new abstraction cross-cutting the whole compiler stack.

> `map: Map_Builder` and it is just a struct, which behaves like an empty `Map`, but if you provide it with an argument, it should consume it and add it to the underlying map.

I guess it was this sentence which inspired me to write the [Map.make <| "A" <- 3 <| "B" <- 5](https://github.com/orgs/enso-org/discussions/6167#discussioncomment-5585150) proposal.

**Reply by kazcw** [reply]
> When one is supposed to treat `ArgumentBlockApplication` as set of named arguments and when one is supposed to treat it as a sequence of statements?

In the tree, this is distinguished by the block type: `ArgumentBlockApplication` is always application of arguments--they can be mixed named- and unnamed-arguments. A sequence of statements is represented as a `BodyBlock`. Syntactically, the distinction is: If a block is introduced by a line ending in an operator (or certain keywords, as in `if`/`else`), it is parsed as a *body block*; otherwise, it is an *argument block* (unless it fits the *operator block* syntax).

**Comment by JaroslavTulach** [comment]
If we shall go with the _"block based syntax"_, then we don't need any parser changes at all. This diff implements such syntax fully in Enso standard libraries:
```diff
enso$ git diff
diff --git distribution/lib/Standard/Base/0.0.0-dev/src/Data/Map.enso distribution/lib/Standard/Base/0.0.0-dev/src/Data/Map.enso
index ad0a4a85f1..f684e1c88b 100644
--- distribution/lib/Standard/Base/0.0.0-dev/src/Data/Map.enso
+++ distribution/lib/Standard/Base/0.0.0-dev/src/Data/Map.enso
@@ -5,6 +5,8 @@ import project.Data.Text.Extensions
 import project.Data.Text.Text
 import project.Errors.Illegal_Argument.Illegal_Argument
 import project.Errors.No_Such_Key.No_Such_Key
+import project.Runtime.State
+import project.Runtime.Ref.Ref
 
 from project.Data.Boolean import Boolean, True, False
 from project import Error, Nothing, Any, Panic
@@ -23,6 +25,12 @@ from project import Error, Nothing, Any, Panic
    treated as a Map.
 @Builtin_Type
 type Map key value
+    make : Any -> Map
+    make ~action =
+        ref = Ref.new Map.empty
+        State.run Map ref action
+        ref.get
+
     ## Returns an empty map.
     empty : Map
     empty = @Builtin_Method "Map.empty"
@@ -357,3 +365,9 @@ type Map key value
     ## PRIVATE
     get_builtin : Any -> Any -> Any
     get_builtin self key ~if_missing = @Builtin_Method "Map.get_builtin"
+
+Any.$ self that =
+    ref = State.get Map
+    map = ref.get
+    new_map = map.insert self that
+    ref.put new_map
diff --git test/Tests/src/Data/Map_Spec.enso test/Tests/src/Data/Map_Spec.enso
index 64975c0676..3b3f84348f 100644
--- test/Tests/src/Data/Map_Spec.enso
+++ test/Tests/src/Data/Map_Spec.enso
@@ -359,6 +359,13 @@ spec =
             m2.to_vector.sort on=_.first . should_equal [["A", 1], ["B", 2]]
             m3.to_vector.sort on=_.first . should_equal [["A", 1], ["C", 3]]
 
+        Test.specify "nicer syntax for maps" <|
+            m1 = Map.make <|
+                "A" $ 1
+                "B" $ 2
+                "C" $ 3
+            m1.to_vector.sort on=_.first . should_equal [["A", 1], ["B", 2], ["C", 3]]
+
         Test.specify "should handle inserts with same keys (1)" <|
             m1 = Map.singleton "A" 1
             m2 = m1.insert "A" 2
```
one can probably use other symbol instead of `$`. I tried `:`, but that one wasn't working. Usage of `=` is also out of question, but others (like `<<=` or `<-`) seem to be fine.

**Reply by wdanilo** [reply]
This is cool, but TBH:

```
+            m1 = Map.make <|
+                "A" $ 1
+                "B" $ 2
+                "C" $ 3
```

Looks like a hack, not a clean solution we should ship to users. It's cool though :)

**Comment by jdunkerley** [comment]
Expanding on this ticket and bring it back to life...

I think the multiline map syntax, would not work well within the node editing style we have (though would with a vertical editor like proposed for Vector). The brackets and inline style would be clear for users and render simply in the ide.

I think @wdanilo simplified `{<key>=<value>, ...}` is the most familiar to users.
I have no objections to also supporting the multi-line style no brackets literal for code based development, but feel the bracket inline style is a high priority literal to add.

cc: @sylwiabr

**Reply by wdanilo** [reply]
@jdunkerley I never proposed multiline-only syntax. Please note that

```
v1 = map
    a = 1
    b = 2
    c = 3
```

is the same as writing

```
v = map a=1 b=2 c=3
```

However, if you think that `{..., ...}` syntax is better, we can use it. In fact, we already have syntax `[..., ...]`, which is not white-space based, so maybe it's not as bad as I initially thought! :)
